pytest_datadir/plugin.py

```diff
@@ -25,24 +26,106 @@ def _win32_longpath(path):
         return path
 
 
-@pytest.fixture
-def shared_datadir(request, tmpdir):
-    original_shared_path = os.path.join(request.fspath.dirname, 'data')
-    temp_path = Path(str(tmpdir.join('data')))
-    shutil.copytree(_win32_longpath(original_shared_path), _win32_longpath(str(temp_path)))
-    return temp_path
-
-
 @pytest.fixture
 def original_datadir(request):
     return Path(os.path.splitext(request.module.__file__)[0])
 
 
-@pytest.fixture
-def datadir(original_datadir, tmpdir):
-    result = Path(str(tmpdir.join(original_datadir.stem)))
-    if original_datadir.is_dir():
-        shutil.copytree(_win32_longpath(str(original_datadir)), _win32_longpath(str(result)))
-    else:
-        result.mkdir()
-    return result
+class DatadirFactory(object):
+    """Factory class for generating datadir fixtures."""
+
+    def __init__(self, request, tmp_path_factory):
+
+        self.tmp_path_factory = tmp_path_factory
+        self.request = request
+
+    def mkdatadir(self, original_datadir='data'):
+
+        # special condition if the datadir is specified as None, which
+        # automatically gets the path that matches the basename of the
+        # module we are in
+        if original_datadir is None:
+            original_datadir = Path(os.path.splitext(self.request.module.__file__)[0])
+
+
+        # get the path to the shared data dir
+        original_path = Path(self.request.fspath.dirname) / original_datadir
+
+        # make sure that the path exists and it is a directory
+        if not original_path.exists():
+            raise ValueError("datadir path does not exist")
+        elif not original_path.is_dir():
+            raise ValueError("datadir path is not a directory")
+
+        # generate a base temporary directory and receive the path to it
+        temp_path = self.tmp_path_factory.mktemp('')
+
+        # in order to use the shutil.copytree util the target directory
+        # must not exist so we specify a dir in the generated tempdir for it
+        temp_data_path = temp_path / 'data'
+
+        # windows-ify the paths
+        original_path = Path(_win32_longpath(original_path))
+        temp_data_path = Path(_win32_longpath(str(temp_data_path)))
+
+        # copy all the files in the original data dir to the temp
+        # dir
+        shutil.copytree(original_path, temp_data_path)
+
+        return temp_data_path
+
+# scoped factories
+@pytest.fixture(scope='module')
+def module_datadir_factory(request, tmp_path_factory):
+
+    return DatadirFactory(request, tmp_path_factory)
```

Comment:
`tmp_path_factory` is session-scoped, and all fixtures are using it even though they have different scopes. This is fine, but this means the directories won't be removed at the end of their respective scopes (`module_datadir_factory` for example won't remove its directory at the end of its module scope).

Is that what you intended?

Reply:
Yes I think this is an issue which could cause considerable confusion. What do you think of this: https://github.com/salotz/pytest-datadir/blob/scoped_tmp_path_factory/pytest_datadir/plugin.py#L81
Could almost be pushed upstream...

Reply:
Merged that proposal for solving this.